stanify/builders/stan_model.py

```diff
@@ -0,0 +1,390 @@
+        lhs_variables = [node.id for node in ast.walk(ast.parse(lhs_expr)) if isinstance(node, ast.Name)]
+        assert len(lhs_variables) == 1, "The LHS expression for prior specification must contain just 1 parameter name"
+        self.lhs_variable = lhs_variables[0]
+        self.distribution_type = distribution_type
+        self.distribution_args = tuple(str(x) for x in distribution_args)
+        self.rhs_variables = set()
+        self.rhs_variables.update([node.id for arg in self.distribution_args for node in ast.walk(ast.parse(str(arg))) if isinstance(node, ast.Name)])
+        self.lower = lower
+        self.upper = upper
+        self.init_state = init_state
+        self.assignment_dist = "assignment"
+
+    def __post_init__(self):
+        if self.distribution_type in ("bernoulli", "binomial", "beta_binomial", "neg_binomial", "poisson"):
+            self.distribution_return_type = int
+        else:
+            # TODO: Check if it's a valid stan distribution
+            self.distribution_return_type = float
+
+# dataclass remove boilerplate
+@dataclass
+class StanDataEntry:
+    data_name: str
+    stan_type: str
+
+@dataclass
+class StanModelContext:
+    initial_time: float
+    integration_times: Iterable[float]
+    stan_data: Dict[str, StanDataEntry] = field(default_factory=dict) # dictionary is obj, lambda function (not obj butc lass)
+    sample_statements: List[SamplingStatement] = field(default_factory=list)
+    exposed_parameters: Set[str] = field(default_factory=set)  # stan variables to be passed to the ODE function (include both hier, nonhier)
+    all_stan_variables: Set[str] = field(default_factory=set)  # set of all stan variables
+    target_integ_outcome_vector_names : Set[str] = field(default_factory=set)
+    obs_integ_outcome_vector_names: Set[str] = field(default_factory=set) # vector connecting draws2data, data2draws
+
+
+    def identify_stan_data_types(self, numeric):
+        def get_dims(obj):
+            if isinstance(obj, np.ndarray):
+                return obj.shape
+            try:
+                iter(obj)
+            except:
+                return None
+            else:
+                dim = len(obj)
+                if dim == 1:
+                    return None
+                inner_dim = get_dims(obj[0])
+                if inner_dim:
+                    return [dim] + inner_dim
+                else:
+                    return [dim]
+
+        for key, val in numeric.items():
+            if isinstance(val, int):
+                self.stan_data[key] = StanDataEntry(key, "int")
+            elif isinstance(val, float):
+                self.stan_data[key] = StanDataEntry(key, "real")
+            else:
+                # Multidimensional data
+                dims = get_dims(val)
+                if not dims:
+                    raise Exception(f"Can't process data entry {key}.")
+                elif len(dims) == 1:
+                    # TODO @Dashadower dims[0] VS using Q, R
```

Comment:
> dims[0] VS using Q, R

I'm not sure what `Q, R` means here. The code below finds the shape of nested lists/multidimensional arrays by recursively checking the length of each dimension, assuming it's not a ragged array.

Reply:
Q is the number of timeseries connecting draws2data and data2draws. So data is N by Q. R is subscript range theta[R]. So this q was how generic our model should be; 

theta_loc_prior = .25

theta ~ normal(theta_loc, theta_loc_prior * .1)
theta_loc ~ exp(1/theta_loc_prior)

for .25 mean parameter theta is our first try for generic hier model.